numpysocket/numpysocket.py

```diff
@@ -3,45 +3,55 @@
 from io import BytesIO
 import logging
 import socket
-from typing import Any, Union
+from typing import Any
 
 import numpy as np
 
 
 class NumpySocket(socket.socket):
+
+    prev = None
+
     def sendall(self, frame: np.ndarray) -> None:  # type: ignore[override]
         out = self.__pack_frame(frame)
+
         super().sendall(out)
         logging.debug("frame sent")
 
+
     def recv(self, bufsize: int = 1024) -> np.ndarray:  # type: ignore[override]
         length = None
         frame_buffer = bytearray()
+        frames = []
         while True:
-            data = super().recv(bufsize)
+
+            if self.prev == None or len(self.prev) == 0:
+                data = super().recv(bufsize)
+            else:
+                data = self.prev
+                self.prev = None
+
             if len(data) == 0:
                 return np.array([])
             frame_buffer += data
-            if len(frame_buffer) == length:
-                break
-            while True:
-                if length is None:
-                    if b":" not in frame_buffer:
-                        break
-                    length_str, _, frame_buffer = frame_buffer.partition(b":")
-                    length = int(length_str)
-                if len(frame_buffer) < length:
-                    break
-
-                frame_buffer = frame_buffer[length:]
-                length = None
+
+            if length is None:
+                if b":" not in frame_buffer:
+                    assert(False) # throw error not np array
```

Comment:
no asserts in a library -- if you're going to make this a failure mode, it should throw an exception